Refactor all tests to use safe expect batcher and general prompt expression

alonSadan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

https://github.com/kubevirt/kubevirt/pull/3882 introduced a safer expect-batcher, which reduces the chance of matching
"leftovers" from previous commands, and by that making the expect-batcher
less prone to bugs. Further details and use-cases can be found in the
PR and in the links mentioned in It's description.

This PR refactors all tests in kubevirt/kubevirt to use the safer expect-batcher, And make use of the general prompt expression.

Also, a long the way and some of the tests, so you can find in this PR also: 
 - modifications to migration, configurations pausing  tests.
 - modifications to login.go in the tests package.  

Signed-off-by: alonSadan asadan@redhat.com

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
Each commit in this PR refactors a different test module.  

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

alonSadan commented:
/test pull-kubevirt-e2e-k8s-1.18

alonSadan commented:
/retest

alonSadan commented:
/test pull-kubevirt-e2e-k8s-1.18

alonSadan commented:
/test pull-kubevirt-e2e-k8s-1.18

alonSadan commented:
/test pull-kubevirt-e2e-k8s-1.18

gabrielecerami commented:
@alonSadan: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

alonSadan commented:
/test pull-kubevirt-e2e-k8s-1.18

jean-edouard commented on tests/config_test.go:
```diff
@@ -495,17 +500,20 @@ var _ = Describe("[rfe_id:899][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 					&expect.BExp{R: tests.RetValue("0")},
-					&expect.BSnd{S: "grep \"PRIVATE KEY\" /mnt/ssh-privatekey\n"},
+					&expect.BSnd{S: "grep -c \"PRIVATE KEY\" /mnt/ssh-privatekey\n"},
+					&expect.BExp{R: tests.RetValue("[1-9]\\d*")},
```
Nit: using a raw string would lift the need for escaping the escape character (`` `[1-9]\d*` ``).

alonSadan replied:
Thanks. Done.

jean-edouard reviewed (approved):
Thanks for doing this! So many `"#"`...
Unfortunately this seems to be "breaking" a lot of tests (more likely reveal mistakes in them).
Good luck fixing those, let me know if I can help.

alonSadan commented:
> Thanks for doing this! So many `"#"`...
> Unfortunately this seems to be "breaking" a lot of tests (more likely reveal mistakes in them).
> Good luck fixing those, let me know if I can help.

Thanks @jean-edouard 

I have modified the safe expecter a bit more and added a `\s?` to it , since somewhen wating for prompt it has space and sometimes it doesn't.

also I [changed](https://github.com/kubevirt/kubevirt/compare/fbd5ac776cdd55bb150702b9fc84e0687075e9d7..a3ab948d2f6c3439c775e8c5ad93e458a4bbebfd)  one of the configuration tests, to not expect `sda` disk in the cirros vm. I think this was a mistake in the in test which the old Expecter covered for. Anyway, there were much less failed test on the first change,  let's see what happens now.

alonSadan commented:
/test pull-kubevirt-e2e-k8s-1.18

alonSadan commented:
tests are passing on my local env, and for some reason not passing here. making it draft, and trying some stuff to solve it

alonSadan commented:
/test pull-kubevirt-e2e-k8s-1.18

alonSadan commented:
/retest

alonSadan commented:
retest

alonSadan commented:
/retest

alonSadan commented:
/retest

alonSadan commented:
/retest

alonSadan commented:
/retest

alonSadan commented:
/test pull-kubevirt-e2e-k8s-1.16

alonSadan commented:
/retest

alonSadan commented:
@jean-edouard  we are all green now. Would you take another look? 

I think since your last visit I also modified migration, pause and configurations tests , and also the made some changes to login.go.

alonSadan commented:
@AlonaKaplan  if you can take a look as well cause I made some changes to the sage expecter

jean-edouard commented on tests/expecter.go:
```diff
@@ -228,7 +229,9 @@ func ExpectBatchWithValidatedSend(expecter expect.Expecter, batch []expect.Batch
 			// Remove the \n since it is translated by the console to \r\n.
 			previousSend = strings.TrimSuffix(previousSend, "\n")
-			bExp.R = fmt.Sprintf("%s%s%s", previousSend, "((?s).*)", bExp.R)
+			bExp.R = fmt.Sprintf("%s%s%s|^\\s?%s*%s", previousSend, "((?s).*)", bExp.R, CRLF, bExp.R)
```
Please use a raw string to avoid the `\\`.
As for the regex, I don't understand what this is doing. Maybe consider adding parentheses to make the operands of `|` more explicit.
Probably also deserves a comment.

alonSadan replied:
> Please use a raw string to avoid the `\\`.

Done.
> As for the regex, I don't understand what this is doing. Maybe consider adding parentheses to make the operands of `|` more explicit.

I added the parentheses, though I am not sure this is what you have meant, can you confirm?
> Probably also deserves a comment.

Done.

jean-edouard replied:
I guess I was wondering about the precedence of `|`, but looks like it's low enough to be evaluated last.
What I meant for the parentheses was to make the left and right operands of `|` explicit, but it's literally `(everything on the left) | (everything on the right)`, so not really useful.
The comment helps a lot though, thank you!
@AlonaKaplan please have a quick look too.

jean-edouard commented on tests/expecter.go:
```diff
@@ -228,7 +229,9 @@ func ExpectBatchWithValidatedSend(expecter expect.Expecter, batch []expect.Batch
-			bExp.R = fmt.Sprintf("%s%s%s", previousSend, "((?s).*)", bExp.R)
+			bExp.R = fmt.Sprintf("%s%s%s|^\\s?%s*%s", previousSend, "((?s).*)", bExp.R, CRLF, bExp.R)
+			//avoid \r in the middle of a command
+			strings.ReplaceAll(bExp.R, " ", `( |\\r)+`)
```
Why do you need 2 backslashes here?

alonSadan replied:
I think that this is been complied as a regular epxresison to match `\r` that is the literal `\` and the literal `r`. as this is what apears in the raw console log (example: {"component":"tests","level":"info","msg":"[{1 stty cols \r 500 rows 500\r\n[root@testvmih4q6ltg6m8xqhqsklbzz8gndrnp6nl5q82z87v4mgsg66rwb fedora]#  []}])

as you can see the raw log has literally `\r` written threre. if I'll use just one backslash it will match a carriage return which is something like :

```
carriage
return
```
and not the literal `\r`.
I hope I explained it well

jean-edouard replied:
I believe those logs you quoted actually make carriage returns explicit by replacing them with `\r` `\n`, but I don't think the console actually contain those as literals. Probably worth double-checking.
For example, I wrote the following recently and used the same logs to get the escape codes: https://github.com/kubevirt/kubevirt/blob/93c2d38d99189915662395fcc390678979296661/tests/expecter.go#L153

alonSadan replied:
I understand. I think it failed for me with "\\r" which suppose to be the same as `\r`, but I'll double check.

AlonaKaplan replied:
["stty cols 500 rows 500\n"](https://github.com/kubevirt/kubevirt/blob/1546fe02637980e0a07e42f9c915b3fb4a3857d4/tests/login.go#L188) was added especially to avoid the line wrap scenario.
Can you please attach a link to the tests that were failing on it? Maybe it contained sent commands longer than 500 chars?
BTW, I saw that only fedora adds " \r" on line wrap. Cirros adds ""\r\r\n". So adding " \n" is not a general solution.

The fix should be not using `ExpectBatchWithValidatedSend` on `configureConsole`. This method is responsible to configure thing so that `ExpectBatchWithValidatedSend` will work. We cannot use it before/during we configure those values.

alonSadan replied:
> ["stty cols 500 rows 500\n"](https://github.com/kubevirt/kubevirt/blob/1546fe02637980e0a07e42f9c915b3fb4a3857d4/tests/login.go#L188) was added especially to avoid the line wrap scenario.
> Can you please attach a link to the tests that were failing on it? Maybe it contained sent commands longer than 500 chars?

[here](https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4010/pull-kubevirt-e2e-k8s-1.16/1314173459147788288#1:build-log.txt%3A5181) line 5181 if the link won't lead you to it.

> BTW, I saw that only fedora adds " \r" on line wrap. Cirros adds ""\r\r\n". So adding " \n" is not a general solution.
> 

maybe I can change it to `\s` which matches both `\r` and `\n`
> The fix should be not using `ExpectBatchWithValidatedSend` on `configureConsole`. This method is responsible to configure thing so that `ExpectBatchWithValidatedSend` will work. We cannot use it before/during we configure those values.

maybe with that modification it will work, and then it worth using it? what do you think?

AlonaKaplan replied:
> > ["stty cols 500 rows 500\n"](https://github.com/kubevirt/kubevirt/blob/1546fe02637980e0a07e42f9c915b3fb4a3857d4/tests/login.go#L188) was added especially to avoid the line wrap scenario.
> > Can you please attach a link to the tests that were failing on it? Maybe it contained sent commands longer than 500 chars?
> 
> [here](https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4010/pull-kubevirt-e2e-k8s-1.16/1314173459147788288#1:build-log.txt%3A5181) line 5181 if the link won't lead you to it.
> 

As I suspected, it happened in the `configureConsole` method with is doing `stty cols 500 rows 500\n` to prevent it.
`ExpectBatchWithValidatedSend` shouldn't be used for `configureConsole`.

> > BTW, I saw that only fedora adds " \r" on line wrap. Cirros adds ""\r\r\n". So adding " \n" is not a general solution.
> 
> maybe I can change it to `\s` which matches both `\r` and `\n`
> 

I believe you shouldn't change it, you should remove it (the whole commit). If you won't use `ExpectBatchWithValidatedSend` on `configureConsole` you should see this issue anymore.

> > The fix should be not using `ExpectBatchWithValidatedSend` on `configureConsole`. This method is responsible to configure thing so that `ExpectBatchWithValidatedSend` will work. We cannot use it before/during we configure those values.
> 
> maybe with that modification it will work, and then it worth using it? what do you think?

I think this commit should be removed from the PR:)

alonSadan replied:
OK. removed the second commit

jean-edouard commented on tests/login.go:
```diff
@@ -113,49 +110,28 @@ func LoggedInFedoraExpecter(vmi *v1.VirtualMachineInstance) (expect.Expecter, er
-				T: expect.OK(),
-			},
-		}},
+		&expect.BExp{R: fmt.Sprintf(`(localhost|%s) login: `, vmi.Name)},
```
There's a good reason for the Caser above. Sometimes we try to login too early, and credentials are not yet setup, which leads to test flakyness.

alonSadan replied:
I understand, but I think it didn't work as expected, it matched the last case - the prompt - even though it didn't actually reached the prompt. it found a `$` (or a `#`) in the login log(in the ssh key image for example) , and "matched" though it didn't ever got to the login prompt. so you think there is another way to address this issue without  giving up on the cases?

Another thing to consider is that maybe the flakyness you are talking  was caused by the issue I mentioned above? I ran those tests many times and I don't think I saw the issue you are talking about

jean-edouard replied:
Good catch for the `$ ` matching outside of login, we do need to fix that.
I'm not sure which issue you're referring to, but the reason we need the Caser is that we need to wait until cloud-init is done, which may happen after the login prompt.
Can we keep the Caser and expect the full prompt instead? (`[fedora@vmi-fedora ~]$ `)

alonSadan replied:
> Can we keep the Caser and expect the full prompt instead? (`[fedora@vmi-fedora ~]$ `)

I think that will  work, but will will loose the generalization we hoped to achieve with the prompt expression.

maybe we can check the  /var/lib/cloud/instance/boot-finished file for existence. something like: 

```
		&expect.BSnd{S: "ls /var/lib/cloud/instance/boot-finished\n"},
		&expect.BExp{R: PromptExpression},
		&expect.BSnd{S: "echo $?\n"},
		&expect.BExp{R: RetValue("0")},
```
This file appears when cloud-init finishes boot. I'll have to verify it, but this should work. what do you think?

jean-edouard replied:
Well you have to be logged-in for that no? :)

alonSadan replied:
haha ooops.
Right.
So I'll use the explicit fedora prompt.

alonSadan replied:
Assumed yes and returned the cases with the full prompt expected in the last case.

alonSadan commented:
sometimes in the ceph lane it looks like the changes I made are not applying : 
https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4010/pull-kubevirt-e2e-k8s-1.17-rook-ceph/1313112469299793920#1:build-log.txt%3A4146

I didn't leave any two `\n` sends in a row, in the whole code,  and this log seems to have it. Weird...
It does not happen all the time though

alonSadan commented:
/retest

alonSadan commented:
Returned cases to fedora login, but expecting the full prompt and not just `$`.
Failed test doesn't seem related to the changes.

jean-edouard commented:
/approve

RamLavi commented on tests/expecter.go:
```diff
@@ -228,7 +229,14 @@ func ExpectBatchWithValidatedSend(expecter expect.Expecter, batch []expect.Batch
+			// Plus in some consoles the prompt is read with a preceded white space, and sometimes it doesn't, so we
+			// allow that as well.
+			bExp.R = fmt.Sprintf("(%s%s%s"+`|^\s?`+"%s*%s)", previousSend, "((?s).*)", bExp.R, CRLF, bExp.R)
+			//avoid a literal \r in the middle of a command.
```
can you reword this comment? it may be stale.

alonSadan replied:
Done.

RamLavi commented on tests/expecter.go:
```diff
@@ -228,7 +229,14 @@ func ExpectBatchWithValidatedSend(expecter expect.Expecter, batch []expect.Batch
+			// the beggining of the line.
+			// Plus in some consoles the prompt is read with a preceded white space, and sometimes it doesn't, so we
+			// allow that as well.
+			bExp.R = fmt.Sprintf("(%s%s%s"+`|^\s?`+"%s*%s)", previousSend, "((?s).*)", bExp.R, CRLF, bExp.R)
```
I'm no regex expert, but is there a way to maybe combine the two options to 1 regex expression?

alonSadan replied:
Me neither :) 
Perhaps @jean-edouard  has a better Idea, but that's the best I could think of. What I was tring here is to  attach the previous BSend *o*r find the desired expression only at the beginning of the line\string (when the buffer is flushed before the expected command is getting in the buffer)

RamLavi commented on tests/expecter.go:
```diff
@@ -201,9 +201,10 @@ func NewConsoleExpecter(virtCli kubecli.KubevirtClient, vmi *v1.VirtualMachineIn
 	}, timeout, opts...)
 }
```
typo in commit message `tests, expecter: improve safe epxecter`: 
the title itself: `epxecter`-->`expecter`
and in the body: `alowes`-->`allows`

alonSadan replied:
Done.

RamLavi reviewed (changes requested):
some nits, plus maybe we need to think of more options how to tackle the issue you mention in commit `tests, expecter: improve safe epxecter`

alonSadan commented on tests/config_test.go:
```diff
@@ -101,9 +101,10 @@ var _ = Describe("[Serial][rfe_id:899][crit:medium][vendor:cnv-qe@redhat.com][le
 				Expect(err).ToNot(HaveOccurred())
 				defer expecter.Close()
 
```
fixed.

alonSadan commented on tests/config_test.go:
```diff
@@ -495,17 +500,20 @@ var _ = Describe("[Serial][rfe_id:899][crit:medium][vendor:cnv-qe@redhat.com][le
 					&expect.BSnd{S: "echo $?\n"},
 					&expect.BExp{R: tests.RetValue("0")},
-					&expect.BSnd{S: "grep \"PRIVATE KEY\" /mnt/ssh-privatekey\n"},
+					&expect.BSnd{S: "grep -c \"PRIVATE KEY\" /mnt/ssh-privatekey\n"},
```
Since the safe expecter [does not allow two BSends in a row](https://github.com/kubevirt/kubevirt/blob/master/tests/expecter.go#L210), I had to manipulate the tests and expect something after the grep command. I could have waited for prompt but I thought this will be more accurate to make it return a number greater than zero. 

I will add a note about it in the commit message.

alonSadan commented:
> some nits, plus maybe we need to think of more options how to tackle the issue you mention in commit `tests, expecter: improve safe expecter`

Thanks Ram. Fixed your comments. 

About the regex in the `improve safe expecter`` commit. if it is about the style and readability of the regex, then I agree it is quite ugly, but I couldn't thought of anything better:(
If it is safety you are worried about, then I think that even though we might loose a bit of the strength the safe expecter had  ( though I am not sure, because the right side of the `|` matches only when the  expected string is at the begging of the buffer, so no leftovers should remain.)  We gained applying it to all tests(and even reviled some errors in them), so I think it is yet better then the regular ExpectBatch.  

With that being said, I would love it if we can think of a more safe and good-looking solution.
Maybe @AlonaKaplan can help us:)

alonSadan commented:
/retest

RamLavi commented:
> About the regex in the `improve safe expecter` commit. if it is about the style and readability of the regex, then I agree it is quite ugly, but I couldn't thought of anything better:(
> If it is safety you are worried about, then I think that even though we might loose a bit of the strength the safe expecter had ( though I am not sure, because the right side of the `|` matches only when the expected string is at the begging of the buffer, so no leftovers should remain.) We gained applying it to all tests(and even reviled some errors in them), so I think it is yet better then the regular ExpectBatch.
> 
> With that being said, I would love it if we can think of a more safe and good-looking solution.
> Maybe @AlonaKaplan can help us:)

I'll try to think of how it could be done, but as I said I'm no regex expert :) 
perhaps we would indeed benefit for @AlonaKaplan 's point of view on this

AlonaKaplan commented:
> **What this PR does / why we need it**:
> 
> https://github.com/kubevirt/kubevirt/pull/3882/files#diff-a5a178b4c66835110ceff63c61d28c96

Can you please change the link to https://github.com/kubevirt/kubevirt/pull/3882 ?

AlonaKaplan commented on tests/config_test.go:
```diff
@@ -495,17 +500,20 @@ var _ = Describe("[Serial][rfe_id:899][crit:medium][vendor:cnv-qe@redhat.com][le
 					&expect.BExp{R: tests.RetValue("0")},
-					&expect.BSnd{S: "grep ssh-rsa /mnt/ssh-publickey\n"},
+					&expect.BSnd{S: "grep -c ssh-rsa /mnt/ssh-publickey\n"},
+					&expect.BExp{R: tests.RetValue(`[1-9]\d*`)},
```
Since the next line makes sure the `grep` succeeded (`echo $?`). I find the `grep -c` and count check confusing. I would choose one of the three options -
1. Using regular grep and waiting for prompt.
2. Using the grep -c and removing the `echo $?`.
3. 
```
&expect.BSnd{S: "cat /mnt/ssh-publickey\n"},
&expect.BExp{"ssh-rsa"},
```
and removing the `echo $?" part.

AlonaKaplan commented on tests/login.go:
```diff
@@ -45,30 +45,29 @@ func LoggedInCirrosExpecter(vmi *v1.VirtualMachineInstance) (expect.Expecter, er
-		&expect.BExp{R: "\\$"}})
-	resp, err := expecter.ExpectBatch(b, 180*time.Second)
+		&expect.BExp{R: PromptExpression}})
+	resp, err := ExpectBatchWithValidatedSend(expecter, b, 180*time.Second)
```
I wouldn't change the login methods to use the `ExpectBatchWithValidatedSend`.
The login batch starts with searching a text that is already in the expecter buffer. Since `ExpectBatchWithValidatedSend` assuming the expected expression is the output of the sent command, it is not suitable in this case.

AlonaKaplan commented on tests/login.go:
```diff
@@ -82,19 +81,18 @@ func LoggedInAlpineExpecter(vmi *v1.VirtualMachineInstance) (expect.Expecter, er
-		&expect.BExp{R: "localhost:~\\#"}})
-	res, err := expecter.ExpectBatch(b, 180*time.Second)
+		&expect.BExp{R: PromptExpression}})
+	res, err := ExpectBatchWithValidatedSend(expecter, b, 180*time.Second)
```
same

AlonaKaplan commented on tests/storage_test.go:
```diff
@@ -381,12 +381,14 @@ var _ = Describe("[Serial]Storage", func() {
 					&expect.BSnd{S: "echo content > /mnt/checkpoint\n"},
+					&expect.BExp{R: tests.PromptExpression},
 					// The QEMU process will be killed, therefore the write must be flushed to the disk.
 					&expect.BSnd{S: "sync\n"},
```
I would add here `&expect.BExp{R: tests.PromptExpression},`

AlonaKaplan commented on tests/migration_test.go:
```diff
@@ -654,7 +654,7 @@ var _ = Describe("[Serial][rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][leve
 				By("Set wrong time on the guest")
-				_, err = expecter.ExpectBatch([]expect.Batcher{
+				_, err = tests.ExpectBatchWithValidatedSend(expecter, []expect.Batcher{
 					&expect.BSnd{S: "date +%T -s 23:26:00\n"},
```
I would add here &expect.BExp{R: tests.PromptExpression},

AlonaKaplan commented on tests/vm_test.go:
```diff
@@ -940,7 +940,7 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
 					By("Issuing a poweroff command from inside VM")
-					_, err = expecter.ExpectBatch([]expect.Batcher{
+					_, err = tests.ExpectBatchWithValidatedSend(expecter, []expect.Batcher{
 						&expect.BSnd{S: "sudo poweroff\n"},
```
I would add here &expect.BExp{R: tests.PromptExpression},

AlonaKaplan commented on tests/vm_test.go:
```diff
@@ -1097,7 +1097,7 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
 					By("Issuing a poweroff command from inside VM")
-					_, err = expecter.ExpectBatch([]expect.Batcher{
+					_, err = tests.ExpectBatchWithValidatedSend(expecter, []expect.Batcher{
 						&expect.BSnd{S: "sudo poweroff\n"},
```
I would add here &expect.BExp{R: tests.PromptExpression},

AlonaKaplan commented on tests/vm_test.go:
```diff
@@ -1337,7 +1337,7 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
 					By("Issuing a poweroff command from inside VM")
-					_, err = expecter.ExpectBatch([]expect.Batcher{
+					_, err = tests.ExpectBatchWithValidatedSend(expecter, []expect.Batcher{
 						&expect.BSnd{S: "sudo poweroff\n"},
```
same

AlonaKaplan commented on tests/vmi_cloudinit_test.go:
```diff
@@ -87,19 +87,19 @@ var _ = Describe("[rfe_id:151][crit:high][vendor:cnv-qe@redhat.com][level:compon
 
 			By("Checking that the VirtualMachineInstance serial console output equals to expected one")
-			resp, err := expecter.ExpectBatch(commands, timeout)
+			resp, err := tests.ExpectBatchWithValidatedSend(expecter, commands, timeout)
```
Same here as in the login. I don't think that `tests.ExpectBatchWithValidatedSend` is suitable in this case since the expected expression is not a result of the sent command.

AlonaKaplan commented on tests/vmi_configuration_test.go:
```diff
@@ -270,7 +270,7 @@ var _ = Describe("Configurations", func() {
 				By("Checking the number of sockets under guest OS")
-				_, err = expecter.ExpectBatch([]expect.Batcher{
+				_, err = tests.ExpectBatchWithValidatedSend(expecter, []expect.Batcher{
 					&expect.BSnd{S: "grep '^physical id' /proc/cpuinfo | uniq | wc -l\n"},
```
`tests.RetVal` is more suitable here:) But can be done in another PR...

AlonaKaplan commented on tests/vmi_configuration_test.go:
```diff
@@ -297,7 +297,7 @@ var _ = Describe("Configurations", func() {
-				_, err = expecter.ExpectBatch([]expect.Batcher{
+				_, err = tests.ExpectBatchWithValidatedSend(expecter, []expect.Batcher{
 					&expect.BSnd{S: "grep '^physical id' /proc/cpuinfo | uniq | wc -l\n"},
 					&expect.BExp{R: "2"},
```
same

AlonaKaplan commented on tests/vmi_configuration_test.go:
```diff
@@ -326,7 +326,7 @@ var _ = Describe("Configurations", func() {
-				_, err = expecter.ExpectBatch([]expect.Batcher{
+				_, err = tests.ExpectBatchWithValidatedSend(expecter, []expect.Batcher{
 					&expect.BSnd{S: "grep '^physical id' /proc/cpuinfo | uniq | wc -l\n"},
 					&expect.BExp{R: "2"},
```
same

AlonaKaplan commented on tests/vmi_configuration_test.go:
```diff
@@ -356,7 +356,7 @@ var _ = Describe("Configurations", func() {
-				_, err = expecter.ExpectBatch([]expect.Batcher{
+				_, err = tests.ExpectBatchWithValidatedSend(expecter, []expect.Batcher{
 					&expect.BSnd{S: "grep -c ^processor /proc/cpuinfo\n"},
 					&expect.BExp{R: "4"},
```
same

AlonaKaplan commented on tests/vmi_configuration_test.go:
```diff
@@ -567,7 +567,7 @@ var _ = Describe("Configurations", func() {
-				res, err := expecter.ExpectBatch([]expect.Batcher{
+				res, err := tests.ExpectBatchWithValidatedSend(expecter, []expect.Batcher{
 					&expect.BSnd{S: "free -m | grep Mem: | tr -s ' ' | cut -d' ' -f2\n"},
 					&expect.BExp{R: "104"},
```
same

AlonaKaplan commented on tests/vmi_configuration_test.go:
```diff
@@ -592,7 +592,7 @@ var _ = Describe("Configurations", func() {
-				res, err := expecter.ExpectBatch([]expect.Batcher{
+				res, err := tests.ExpectBatchWithValidatedSend(expecter, []expect.Batcher{
 					&expect.BSnd{S: "free -m | grep Mem: | tr -s ' ' | cut -d' ' -f2\n"},
 					&expect.BExp{R: "104"},
```
same

AlonaKaplan commented on tests/vmi_configuration_test.go:
```diff
@@ -811,7 +811,7 @@ var _ = Describe("Configurations", func() {
-				_, err = expecter.ExpectBatch([]expect.Batcher{
+				_, err = tests.ExpectBatchWithValidatedSend(expecter, []expect.Batcher{
 					&expect.BSnd{S: "grep -rs '^QEMU Virtio Tablet' /sys/devices | wc -l\n"},
 					&expect.BExp{R: "1"},
```
same

AlonaKaplan commented on tests/vmi_configuration_test.go:
```diff
@@ -838,7 +838,7 @@ var _ = Describe("Configurations", func() {
-				_, err = expecter.ExpectBatch([]expect.Batcher{
+				_, err = tests.ExpectBatchWithValidatedSend(expecter, []expect.Batcher{
 					&expect.BSnd{S: "grep -rs '^QEMU USB Tablet' /sys/devices | wc -l\n"},
 					&expect.BExp{R: "1"},
```
same

alonSadan replied:
Done.

AlonaKaplan commented on tests/vmi_configuration_test.go:
```diff
@@ -1149,7 +1149,7 @@ var _ = Describe("Configurations", func() {
-				_, err = expecter.ExpectBatch([]expect.Batcher{
+				_, err = tests.ExpectBatchWithValidatedSend(expecter, []expect.Batcher{
 					&expect.BSnd{S: "grep -c ^virtio /sys/devices/virtual/misc/hw_random/rng_available\n"},
 					&expect.BExp{R: "1"},
```
same

alonSadan replied:
Done.

AlonaKaplan commented on tests/vmi_configuration_test.go:
```diff
@@ -1168,7 +1168,7 @@ var _ = Describe("Configurations", func() {
-				_, err = expecter.ExpectBatch([]expect.Batcher{
+				_, err = tests.ExpectBatchWithValidatedSend(expecter, []expect.Batcher{
 					&expect.BSnd{S: "[[ ! -e /sys/devices/virtual/misc/hw_random/rng_available ]] && echo non\n"},
 					&expect.BExp{R: "non"},
```
same

alonSadan replied:
Done.

AlonaKaplan commented on tests/vmi_configuration_test.go:
```diff
@@ -1253,7 +1253,7 @@ var _ = Describe("Configurations", func() {
 				By("Terminating guest agent and waiting for it to dissappear.")
-				res, err := expecter.ExpectBatch([]expect.Batcher{
+				res, err := tests.ExpectBatchWithValidatedSend(expecter, []expect.Batcher{
 					&expect.BSnd{S: "systemctl stop guestagent\n"},
```
I would add `Exp` for prompt.

alonSadan replied:
Done.

AlonaKaplan commented on tests/vmi_configuration_test.go:
```diff
@@ -1351,7 +1351,7 @@ var _ = Describe("Configurations", func() {
 				By("Terminating guest agent and waiting for it to dissappear.")
-				res, err := expecter.ExpectBatch([]expect.Batcher{
+				res, err := tests.ExpectBatchWithValidatedSend(expecter, []expect.Batcher{
 					&expect.BSnd{S: "systemctl stop guestagent\n"},
```
I would add `Exp` for prompt.

alonSadan replied:
Done.

AlonaKaplan commented on tests/vmi_configuration_test.go:
```diff
@@ -1589,7 +1589,7 @@ var _ = Describe("Configurations", func() {
-				_, err = expecter.ExpectBatch([]expect.Batcher{
+				_, err = tests.ExpectBatchWithValidatedSend(expecter, []expect.Batcher{
 					&expect.BSnd{S: fmt.Sprintf("grep %s /proc/cpuinfo\n", vmiModel)},
 					&expect.BExp{R: "model name"},
```
I would use `RetVal`.

AlonaKaplan commented on tests/vmi_ignition_test.go:
```diff
@@ -63,7 +63,7 @@ var _ = Describe("[Serial][rfe_id:151][crit:high][vendor:cnv-qe@redhat.com][leve
 
 			By("Checking that the VirtualMachineInstance serial console output equals to expected one")
-			resp, err := expecter.ExpectBatch(commands, timeout)
+			resp, err := tests.ExpectBatchWithValidatedSend(expecter, commands, timeout)
```
Same as the login. `tests.ExpectBatchWithValidatedSend` is not suitable here since the expected expression is not the output of the sent command.

alonSadan replied:
I think that [here](https://github.com/kubevirt/kubevirt/blob/eee4d2550a6b4e150b1bf0f6de11b07cb1230e0d/tests/vmi_ignition_test.go#L90) we do use it with expected command as a result of a BSend.

AlonaKaplan commented on tests/expecter.go:
```diff
@@ -228,7 +229,16 @@ func ExpectBatchWithValidatedSend(expecter expect.Expecter, batch []expect.Batch
+			// the beggining of the line.
+			// Plus in some consoles the prompt is read with a preceded white space, and sometimes it doesn't, so we
+			// allow that as well.
+			bExp.R = fmt.Sprintf("(%s%s%s"+`|^\s?`+"%s*%s)", previousSend, "((?s).*)", bExp.R, CRLF, bExp.R)
```
This expression breaks the whole idea of `ExpectBatchWithValidatedSend`.
The idea is to make sure the expected expression is the result of the sent command and not a leftover of the buffer.
With your new addition to the expression we rollback to the previous situation where we could find the expected expression somewhere in the buffer left over.

alonSadan replied:
I am not sure it completely breaks it. it may weakens it a bit, but we gained using it also in login batches. the point is to avoid leftovers from previous commands, if we are matching the expression only if is at the beginning of the buffer (with `^` ) we should almost completely avoid matching leftovers, don't we?

I thought of scenario were what I suggested might fail but I think it is really rare and perhaps worth the risk: 
```
					&expect.BSnd{S: "sudo su -\n"},
					&expect.BExp{R: tests.PromptExpression},
					&expect.BSnd{S: "mount /dev/sda /mnt\n"},
					&expect.BExp{R: tests.PromptExpression},
```
Here for example  after we send `sudo su \n` we expect a prompt, and on the second time we expect a prompt we might, somehow, match the first prompt, but it will happen only if the first prompt is appearing right at the beginning of the buffer with nothing besides white-spaces, `\r` and `\n` preceding it. 
if, for example, `sudo su \n` will precede it, the right side of the `|` won't match, and we are back at our safe zone.

make sense, or I'm talking BS?

AlonaKaplan replied:
yes, the buffer leftover may start with the expected result (we have pretty high possibly for it to happen when the expected result if the prompt).
Anyway, the login method start with sending `\n` and then expecting for an expression that is not the result of `\n`. Therefore, those methods are not suitable to be used by the `ExpectBatchWithValidatedSend`.

We have only 3 login method. And you use `ExpectBatchWithValidatedSend` for only two of them. It doesn't worth adapting `ExpectBatchWithValidatedSend` just for those 2 methods. We should make sure those ,methods are safe without `ExpectBatchWithValidatedSend`.

alonSadan replied:
OK then. I will drop the second commit

AlonaKaplan commented on tests/expecter.go:
```diff
@@ -201,9 +201,10 @@ func NewConsoleExpecter(virtCli kubecli.KubevirtClient, vmi *v1.VirtualMachineIn
+// ExpectBatchWithValidatedSend adds the expect.BSnd command to the expect.BExp expression.
 // It is done to make sure the match was found in the result of the expect.BSnd
 // command and not in a leftover that wasn't removed from the buffer.
+// It also expects the Bexp to be at the begining of the line with only CRLF expression preceding it.
```
I don't understand the comment. How can `BExp` be in the beginning of the line? What line? Do you mean at the begining of the batch?

Anyway, I don't understand the commit message as well. Can you please provide an example?
Why the buffer go flushed?

I'm not sure since I didn't understand this commit.
But since I saw in the commit message the problem (which I don't understand:)) is mostly on login I just want to mention as I mentioned in the previous commit. `ExpectBatchWithValidatedSend` is not suitable for login methods.
The main idea behind `ExpectBatchWithValidatedSend` is that it makes sure the expected expression is the result of the sent command.
The method is not suitable for batches as the login where we first send `\n` but the expression we are looking for may appear in the buffer much before we sent the `\n`.

AlonaKaplan replied:
I don't want to lose a bit of the safe expecter for two login methods. The logion methods are not suitable for `ExpectBatchWithValidatedSend` as I explained in the previous comment.

AlonaKaplan commented on tests/login.go:
```diff
@@ -179,21 +177,21 @@ func ReLoggedInFedoraExpecter(vmi *v1.VirtualMachineInstance, timeout int) (expe
 		&expect.BSnd{S: "echo $?\n"},
 		&expect.BExp{R: RetValue("0")}})
-	resp, err := expecter.ExpectBatch(batch, 30*time.Second)
+	resp, err := ExpectBatchWithValidatedSend(expecter, batch, 30*time.Second)
```
`configureConsole` shouldn't use `ExpectBatchWithValidatedSend`. `configureConsole` config things so that is will be possible to use `ExpectBatchWithValidatedSend`. We cannot use `ExpectBatchWithValidatedSend` before/during the configuration stage.

AlonaKaplan replied:
If you can, I prefer the second commit (tests, expecter: add changes to safe expecter) will be removed from the PR.

AlonaKaplan commented:
/hold Thanks for working on this follow-up PR. Great work. Just please, although the PR is approved, don't merge it as it is. IMO it contains a commit that breaks the whole concept of `ExpectBatchWithValidatedSend`. Please refer to my comment before cancelling the hold.

alonSadan commented:
[fixes](https://github.com/kubevirt/kubevirt/compare/2783693d3d11c9eb762ae04300f60005a9ef26f4..d32146bf2d2a1ec8daebb7d7af9e138158ad0c37) to (most of) @AlonaKaplan 's comments:)

alonSadan commented:
[rebased on master](https://github.com/kubevirt/kubevirt/compare/d32146bf2d2a1ec8daebb7d7af9e138158ad0c37..357e2e5bfaaaef12c704d61ca1a581efc69990ce)

alonSadan commented:
[removed](https://github.com/kubevirt/kubevirt/compare/357e2e5bfaaaef12c704d61ca1a581efc69990ce..8bd6a03a1e69d628ab19af2cf933b9b782b6387b) second commit.

alonSadan commented:
> /hold Thanks for working on this follow-up PR. Great work. Just please, although the PR is approved, don't merge it as it is. IMO it contains a commit that breaks the whole concept of `ExpectBatchWithValidatedSend`. Please refer to my comments before cancelling the hold.

Thanks @AlonaKaplan for reviewing. removed second commit and responded to the comments.

alonSadan commented:
[removed](https://github.com/kubevirt/kubevirt/compare/5f3ce0c88da07af5b6432bda4770b86f0e014fbe..e01ad69f101581704fad758bf05b95d650d63072) safe expecter from login batches

alonSadan commented:
/retest

AlonaKaplan commented:
/lgtm
/hold cancel
Thanks!

alonSadan commented:
/test pull-kubevirt-e2e-k8s-1.17-rook-ceph

dankenigsberg commented on tests/pausing_test.go:
```diff
@@ -373,7 +373,7 @@ var _ = Describe("[rfe_id:3064][crit:medium][vendor:cnv-qe@redhat.com][level:com
 
 		grepSleepPid := func(expecter expect.Expecter) string {
 			res, err := console.ExpectBatchWithValidatedSend(expecter, []expect.Batcher{
-				&expect.BSnd{S: `pgrep -f "sleep 8"` + "\n"},
```
I'm not sure how useful this test originally was, but I think that this change made it worse as it makes the following statement untrue.
```
			By("Waiting longer than the process normally runs")
			time.Sleep(10 * time.Second)
```
Adding more sleep time to solve a race is a known fallacy. You should never do it without explaining why it only sweeps the problem under the rug rather than resolve it.

I think that a better test for pause would be to measure the difference in uptime between guest and host, pause, and measure again. It is still raceful but quicker. To avoid the race you can trust libvirt and simply ask it if the domain is paused.

alonSadan replied:
> I'm not sure how useful this test originally was, but I think that this change made it worse as it makes the following statement untrue.
> 
> ```
> 			By("Waiting longer than the process normally runs")
> 			time.Sleep(10 * time.Second)
> ```
> 
Thank you @dankenigsberg  for this catch, I totally missed it. https://github.com/kubevirt/kubevirt/pull/4394 should quick-fix it.
> Adding more sleep time to solve a race is a known fallacy. You should never do it without explaining why it only sweeps the problem under the rug rather than resolve it.

I agree, but I think that in this case it is not so bad:
The point of this test, as I understand it,  is to check if a process is continued after we pause the vmi. I don't think that testing the pausing action itself is in the scope of this test case.
`sleep` is just been used here as a long process, it could be any other process that runs for a long time.
anyway, It is not causing any other component to wait while something else happening, which might lead to unexpected behavior. (as sometimes happen when using `Eventually` too widely, for example).

The point of this change is to make sure the process doesn't finish before we pause the vmi. so by making it longer we can be more positive about it won't happen. I don't think it sweeps the problem under the rug in that case.
make sense, or am I missing something?

> 
> I think that a better test for pause would be to measure the difference in uptime between guest and host, pause, and measure again. It is still raceful but quicker. To avoid the race you can trust libvirt and simply ask it if the domain is paused.

Sounds good!. I can try and add a test  case to measure that, but I don't think it belongs in this `It` scope.
As for asking libvirt, I think this is done by [retrieving the condition](https://github.com/kubevirt/kubevirt/blob/master/tests/pausing_test.go#L422), which is eventually is [updating from the domain](https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-handler/vm.go#L822). 

What do you think?

dankenigsberg replied:
"we can be more positive" is not "we are certain". The difference between these two statements is what I mean by "sweeping the problem under the rug". You don't solve the problem, you just make it less likely to show up.

rmohr replied:
@alonSadan if you look at https://github.com/kubevirt/kubevirt/pull/4010/commits/cc62349d370c39e4fc5776a12cbd3e1436b4c32b#diff-777ec290af00beeee5968636afc39636fc0399afd3818df5f6efba35bce80419R425 you will see that with the timeout increase the test does actually not test anymore what it is supposed to test. The test waits for 10 seconds with the assumption that the 1m process will have a regular runtime of less than 10 seconds.

@dankenigsberg I agree that this needs another mechanism to test things. However we tend to test, if possible, always end-to-end, so only asking libvirt for the status is not something I would call optimal. Taking uptime or something, like you suggested, sounds like a good plan.

alonSadan replied:
> @alonSadan if you look at cc62349#diff-777ec290af00beeee5968636afc39636fc0399afd3818df5f6efba35bce80419R425 you will see that with the timeout increase the test does actually not test anymore what it is supposed to test. The test waits for 10 seconds with the assumption that the 1m process will have a regular runtime of less than 10 seconds.


@rmohr  yes, I totally missed those two waiting lines. I suggested a quick-fix #4394.

> we can be more positive" is not "we are certain". The difference between these two statements is what I mean by "sweeping the problem under the rug". You don't solve the problem, you just make it less likely to show up

@dankenigsberg  I understand. My reason was to apply the safe expecter with changing as less as possible the existing tests. so I tried to keep the test mechanism, but make it pass with the safe expecter. This test didn't worked , and the old expecter coverd for it. (after  #4394. will get in) The test will still not be perfect, but at least better then it was.
to improve the tests mechanism I would suggest a different PR.

alonSadan replied:
@dankenigsberg @rmohr 
Eventually I removed the old test and added a new one based on the uptime diff of the guest and the host.
you can see it here  #4394.
Thanks for the help.
